airflow/jobs/scheduler_job.py

```diff
@@ -858,7 +865,7 @@ def _do_scheduling(self, session) -> int:
                     self.log.error("DAG '%s' not found in serialized_dag table", dag_run.dag_id)
                     continue
 
-                self._send_dag_callbacks_to_processor(dag, callback_to_run)
+                callbacks_to_send.append((dag, callback_to_run))
```

Comment:
Hmm, this may change the behaviour in some cases. Judging from the `expunge_all()` call below, I suspect the callbacks are allowed to update the database before the critical section below. Moving the callbacks to after the section means they no longer have this ability.

Reply:
Alternatively we could end the previous guard just before, send callbacks to DB then start another guard below. 
I don't know how important it is to keep the single guard for all these lines 843-900

Reply:
Yeah make sense Jarek:

```python
        with prohibit_commit(session) as guard:
            ...
            guard.commit()

        # Send the callbacks after we commit to ensure the context is up to date when it gets run
        for dag_run, callback_to_run in callback_tuples:
            dag = self.dagbag.get_dag(dag_run.dag_id, session=session)
            if not dag:
                self.log.error("DAG '%s' not found in serialized_dag table", dag_run.dag_id)
                continue

            self._send_dag_callbacks_to_processor(dag, callback_to_run)

        with prohibit_commit(session) as guard:
```

Reply:
Thank you!